README.md
# pistop

I use a Raspberry PI4 to provide NFSv4 and Samba shares to share my Music Directory to all clients. These clients are mostly Single Board Computers (SBCs), mix of Raspbery PIs, Banana PIs, Allo sparky SBC. All of these mount the filesystem using automount. Music playback is done via Music Player Daemon. These SBCs have been installed in each of my room (Living Room, Music Room, two bedrooms and one device which is connected to a Headphone amp. The power supply of these boards are connected to WiFI switches which can be remotely turned off.

The problem **pistop** solves is a synchronized mounting of the shared Music Directory, starting of mpd and automount service as soon as the PI4 server is powered on and up. This is achieved by having a service known as **fclient** on the clients and a service known as **fserver** on the PI4. These services use [supervise](https://en.wikipedia.org/wiki/Daemontools), **tcpclient**, **tcpserver** from the [ucspi-tcp](https://cr.yp.to/ucspi-tcp.html) package. I use the `fclient` service on my laptop too. It automatically stars the mpd daemon, the moment it sees my PI4 server up running the `fserver` service. In case, I'm done with listening to Music, I shutdown the PI4 server. The `fclient` service detects that and stops the automount service, unmounts the NFS share and stops the mpd daemon. This prevents programs like `ls` from hanging when doing a listing of your home directory.

So when I retire for the night to get a good night sleep, I use an android app known as `ssh button` to shutdown the server. The moment the server shuts down, all my clients shutdown together. Using home automation, I then switch off the power output to each and very adaptor thereby saving around 0.05 kwhr of electricity per device every day. You could say that these SBC consume miniscule power, but this kind of philosophy has led us mankind to plunder this planet. Global Warming is a stark reality of the day and we should try to conserve electricity as much as possible.

## INSTALLATION

On the server (use the IP Address of your own server)

```
# apt-get update
# apt-get install pistop
# /usr/libexec/pistop/create_service --servicedir=/service --service_name=fserver --host=192.168.1.1 --port=5555 --add-service
```

On each client (use IP Address of your server)

```
# apt-get update
# apt-get install pistop
# /usr/libexec/pistop/create_service --servicedir=/service --service_name=fclient --host=192.168.1.1 --port=5555 --add-service
```

If you want to have your  clients shutdown itself when the server is powered off, create `POWER_OFF` enviroment variable by doing this
`sudo sh -c "echo 1 > /service/fclient/variables/POWER_OFF"`

## Prebuilt Binaries

Prebuilt binaries using openSUSE Build Service are available for pistop for

* Debian (including ARM images for Debian 10 which work (and tested) for RaspberryPI (model 2,3 & 4) and Allo Sparky)
* Ubuntu

Use the below url for installation

https://software.opensuse.org//download.html?project=home%3Ambhangui%3Adietpi&package=pistop


ChangeLog
*
Release 1.0 Start 28/08/2020
28/08/2020
01. Initial Release
02. fixed syntax error in create_service
03. terminate cat command on SIGTERM
29/08/2020
04. fixed shutdown script
31/08/2020
05. Added code to do backup in fserver
06. redo logic to kill background jobs
07. fixed path for pistop
08. added backup script
25/02/2021
09. service run file - use relative path for variables to allow renaming of
    service directory


client.c
/*
 * $Log: $
 */
#include <stdlib.h>
#include <errno.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/select.h>
#include <subfd.h>
#include <scan.h>
#include <stralloc.h>
#include <strerr.h>
#include <error.h>
#include <env.h>
#include <str.h>
#include <sig.h>
#include <fmt.h>
#include <sgetopt.h>

#define FATAL             "client: fatal: "
#define SELECTTIMEOUT     5

char           *usage = "usage: client [-d timeout]";

void
out(char *str)
{
	if (!str || !*str)
		return;
	if (substdio_puts(subfdout, str) == -1)
		strerr_die2sys(111, FATAL, "write: ");
	return;
}

void
flush()
{
	if (substdio_flush(subfdout) == -1)
		strerr_die2sys(111, FATAL, "write: ");
}

void
err(char *str)
{
	if (!str || !*str)
		return;
	if (substdio_puts(subfderr, str) == -1)
		strerr_die2sys(111, FATAL, "write: ");
	return;
}

void
errf(char *str)
{
	if (!str || !*str)
		return;
	if (substdio_puts(subfderr, str) == -1)
		strerr_die2sys(111, FATAL, "write: ");
	else
	if (substdio_flush(subfderr) == -1)
		strerr_die2sys(111, FATAL, "write: ");
	return;
}

void
die_nomem()
{
	errf("out of memory\n");
	_exit(111);
}

void
die_read()
{
	errf("read error\n");
	_exit(111);
}

void
die_write()
{
	errf("write error\n");
	_exit(111);
}

void
die_alarm()
{
	errf("timeout\n");
	_exit(111);
}

void
sigterm()
{
	substdio_flush(subfdout);
	substdio_flush(subfderr);
	_exit(1);
}

int
timeoutread(t, fd, buf, len)
	int             t;
	int             fd;
	char           *buf;
	int             len;
{
	fd_set          rfds;
	struct timeval  tv;

	tv.tv_sec = t;
	tv.tv_usec = 0;

	FD_ZERO(&rfds);
	FD_SET(fd, &rfds);

	if (select(fd + 1, &rfds, (fd_set *) 0, (fd_set *) 0, &tv) == -1)
		return -1;
	if (FD_ISSET(fd, &rfds))
		return read(fd, buf, len);

	errno = error_timeout;
	return -1;
}

int
timeoutwrite(t, fd, buf, len)
	int             t;
	int             fd;
	char           *buf;
	int             len;
{
	fd_set          wfds;
	struct timeval  tv;

	tv.tv_sec = t;
	tv.tv_usec = 0;

	FD_ZERO(&wfds);
	FD_SET(fd, &wfds);

	if (select(fd + 1, (fd_set *) 0, &wfds, (fd_set *) 0, &tv) == -1)
		return -1;
	if (FD_ISSET(fd, &wfds))
		return write(fd, buf, len);

	errno = error_timeout;
	return -1;
}

int
main(int argc, char **argv)
{
	int             opt, length, dataTimeout = -1, r;
	struct timeval  timeout;
	struct timeval *tptr;
	time_t          last_timeout;
	char            buffer[256];
	char           *ptr;
	fd_set          rfds; /*- File descriptor mask for select -*/

	sig_termcatch(sigterm);
	while ((opt = getopt(argc, argv, "d:")) != opteof) {
		switch (opt) {
		case 'd':
			scan_int(optarg, &dataTimeout);
			break;
		}
	}
	if (dataTimeout == -1) {
		if (!(ptr = env_get("DATA_TIMEOUT")))
			ptr = "1800";
		scan_int(ptr, &dataTimeout);
	}
	if ((r = timeoutwrite(dataTimeout, 7, "wait\n", 5)) < 0)
		die_write();
	if (dataTimeout > 0)
		timeout.tv_sec = (dataTimeout > SELECTTIMEOUT) ? dataTimeout : SELECTTIMEOUT;
	else
		timeout.tv_sec = 0 - dataTimeout;
	timeout.tv_usec = 0;
	tptr = (timeout.tv_sec ? &timeout : (struct timeval *) 0);

	last_timeout = timeout.tv_sec;
	for (;;) {
		FD_ZERO(&rfds);
		FD_SET(6, &rfds);
		if ((r = select(7, &rfds, (fd_set *) NULL, (fd_set *) NULL, tptr)) < 0) {
#ifdef ERESTART
			if (errno == EINTR || errno == ERESTART)
#else
			if (errno == EINTR)
#endif
				continue;
		}
		if (!r) { /*- timeout occurred */
			if (dataTimeout > 0) {
				last_timeout += 2 * last_timeout;
				timeout.tv_sec = last_timeout;
			} else {
				timeout.tv_sec = 0 - dataTimeout;
				last_timeout = timeout.tv_sec;
			}
			timeout.tv_usec = 0;
		} else {
			if (dataTimeout > 0)
				last_timeout = timeout.tv_sec = (dataTimeout > SELECTTIMEOUT) ? dataTimeout : SELECTTIMEOUT;
			else
				timeout.tv_sec = 0 - dataTimeout;
			timeout.tv_usec = 0;
		}
		if (FD_ISSET(6, &rfds)) { /*- data from socket/stdin and display it */
			if ((length = read(6, buffer, sizeof(buffer))) == -1)
				strerr_die2sys(111, FATAL, "read-network: ");
			if (!length) {
				out("PI4 powerd off\n");
				flush();
				/*
				execl("/sbin/shutdown", "-h", "now", (char *) 0);
				strerr_die2sys(111, FATAL, "execl: /sbin/shutdown -r now: ");
				*/
			} 
			if (substdio_put(subfderr, buffer, length) == -1)
				die_write();
			if (substdio_flush(subfderr) == -1)
				die_write();
		}
	} /*- for (;;) { */
	_exit(0);
}


server.c
/*
 * $Log: server.c,v $
 * Revision 1.4  2020-08-31 21:18:21+05:30  Cprogrammer
 * fixed compiler warnings
 *
 * Revision 1.3  2020-08-29 17:12:58+05:30  Cprogrammer
 * removed creation of /tmp/dietpiserver.pid
 *
 * Revision 1.2  2020-08-29 16:02:07+05:30  Cprogrammer
 * change client message
 *
 * Revision 1.1  2020-08-28 20:37:41+05:30  Cprogrammer
 * Initial revision
 *
 */
#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif
#ifdef HAVE_ERRNO_H
#include <errno.h>
#endif
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#ifdef HAVE_SYS_SELECT_H
#include <sys/select.h>
#endif
#ifdef HAVE_QMAIL
#include <subfd.h>
#include <scan.h>
#include <open.h>
#include <strerr.h>
#include <env.h>
#include <str.h>
#include <sig.h>
#include <fmt.h>
#include <sgetopt.h>
#endif

#define FATAL             "server: fatal: "
#define SELECTTIMEOUT     5

char           *usage = "usage: server [-d timeout]";

void
out(char *str)
{
	if (!str || !*str)
		return;
	if (substdio_puts(subfdout, str) == -1)
		strerr_die2sys(111, FATAL, "write: ");
	return;
}

void
sigterm()
{
	substdio_flush(subfdout);
	substdio_flush(subfderr);
	_exit(1);
}

#ifdef DEBUG
static char     strnum[FMT_ULONG];
#endif

int
main(int argc, char **argv)
{
	int             opt, length, dataTimeout = -1, retval;
	struct timeval  timeout;
	struct timeval *tptr;
	time_t          last_timeout;
	char            buffer[256];
	char           *ptr;
	fd_set          rfds;	/*- File descriptor mask for select -*/

	sig_termcatch(sigterm);
	while ((opt = getopt(argc, argv, "d:")) != opteof) {
		switch (opt) {
		case 'd':
			scan_int(optarg, &dataTimeout);
			break;
		}
	}
	if (dataTimeout == -1) {
		if (!(ptr = env_get("DATA_TIMEOUT")))
			ptr = "1800";
		scan_int(ptr, &dataTimeout);
	}
	if (dataTimeout > 0)
		timeout.tv_sec = (dataTimeout > SELECTTIMEOUT) ? dataTimeout : SELECTTIMEOUT;
	else
		timeout.tv_sec = 0 - dataTimeout;
	timeout.tv_usec = 0;
	tptr = (timeout.tv_sec ? &timeout : (struct timeval *) 0);

	last_timeout = timeout.tv_sec;
	for (;;) {
		FD_ZERO(&rfds);
		FD_SET(0, &rfds);
		if ((retval = select(1, &rfds, (fd_set *) NULL, (fd_set *) NULL, tptr)) < 0) {
#ifdef ERESTART
			if (errno == EINTR || errno == ERESTART)
#else
			if (errno == EINTR)
#endif
				continue;
		}
		if (!retval) { /*- timeout occurred */
			if (dataTimeout > 0) {
				last_timeout += 2 * last_timeout;
				timeout.tv_sec = last_timeout;
			} else {
				timeout.tv_sec = 0 - dataTimeout;
				last_timeout = timeout.tv_sec;
			}
			timeout.tv_usec = 0;
		} else {
			if (dataTimeout > 0)
				last_timeout = timeout.tv_sec = (dataTimeout > SELECTTIMEOUT) ? dataTimeout : SELECTTIMEOUT;
			else
				timeout.tv_sec = 0 - dataTimeout;
			timeout.tv_usec = 0;
		}
		if (FD_ISSET(0, &rfds)) { /*- data from socket/stdin (client) */
			if ((length = read(0, buffer, sizeof(buffer))) == -1)
				strerr_die2sys(111, FATAL, "read-network: ");
			if (!length) {
				if (substdio_puts(subfderr, "client dropped connection\n") == -1)
					strerr_die2sys(111, FATAL, "write: ");
				if (substdio_flush(subfderr) == -1)
					strerr_die2sys(111, FATAL, "write: ");
				break;
			} 
#ifdef DEBUG
			strnum[fmt_ulong(strnum, length)] = 0;
			out("length = ");
			out(strnum);
			out("\n");
			substdio_flush(subfdout);
#endif
			if (!str_diffn(buffer, "shutdown", 8)) {
				execl("/sbin/shutdown", "-h", "now", (char *) 0);
				strerr_die2sys(111, FATAL, "execl: /sbin/shutdown -h now: ");
			} else
			if (!str_diffn(buffer, "quit", 4)) {
				close(0);
				close(1);
				close(2);
				_exit(0);
			} else
			if (!str_diffn(buffer, "wait", 4))
				continue;
			/*- display data from client */
			if (substdio_put(subfderr, buffer, length) == -1)
				strerr_die2sys(111, FATAL, "write: ");
			if (substdio_flush(subfderr) == -1)
				strerr_die2sys(111, FATAL, "write: ");
		}
	} /*- for (;;) { */
	_exit(0);
}
